src/orange/orangeinp/CsgTreeUtils.cc

```diff
@@ -165,6 +171,273 @@ void simplify(CsgTree* tree, NodeId start)
     }
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Simplify negated joins using De Morgan's law.
+ *
+ * This is required if the tree's logic expression is used with
+ * \c InfixEvaluator as negated joins are not supported.
+ */
+CsgTree transform_negated_joins(CsgTree const& tree)
+{
+    // For each node_id in that set, we'll create a Negated{node_id} node in
+    // the new tree.
+    std::unordered_set<NodeId> new_negated_nodes;
+
+    // We need to insert a negated version of these Joined nodes.
+    // The value is the Node referred to by the key as to avoid an extra
+    // get_if.
+    std::unordered_map<NodeId, Joined const*> negated_join_nodes;
+
+    // Contains the node_id of Negated{Joined}. These don't need to be inserted
+    // in the new tree and its parent can directly point to the transformed
+    // Joined node.
+    std::unordered_map<NodeId, Negated const*> transformed_negated_nodes;
+
+    // A Negated{Joined{}} nodes needs to be mapped to the opposite Joined{}
+    // node. Parents need to point to the new Joined{} node. The old Joined{}
+    // node needs to be kept if it had parents other than the Negated node.
+    std::unordered_set<NodeId> orphaned_join_nodes;
+    std::unordered_set<NodeId> orphaned_negate_nodes;
```

Comment:
Since all the node ids are consecutive, a vector of bools might be more effective (maybe not worrying about performance wise)

Reply:
let me make sure everything works before using a vector of bool 😅